WalletWasabi/Services/WalletService.cs

```diff
@@ -586,41 +586,52 @@ public Node LocalBitcoinCoreNode
 			private set => _localBitcoinCoreNode = value;
 		}
 
+		/// <param name="hash">Block hash of the desired block, represented as a 256 bit integer.</param>
 		/// <exception cref="OperationCanceledException"></exception>
-		public async Task<Block> GetOrDownloadBlockAsync(uint256 hash, CancellationToken cancel)
+		public async Task<Block> FetchBlockAsync(uint256 hash, CancellationToken cancel)
+		{
+			Block block = await GetBlockFromFileAsync(hash, cancel);
+			if (block == null)
+			{
+				block = await DownloadBlockAsync(hash, cancel);
+			}
+			return block;
+		}
+
+		/// <param name="hash">Block hash of the desired block, represented as a 256 bit integer.</param>
+		/// <exception cref="OperationCanceledException"></exception>
+		private async Task<Block> GetBlockFromFileAsync(uint256 hash, CancellationToken cancel)
 		{
 			// Try get the block
+			Block block = null;
 			using (await BlockFolderLock.LockAsync())
 			{
 				var encoder = new HexEncoder();
-				foreach (var filePath in Directory.EnumerateFiles(BlocksFolderPath))
+				var filePath = Path.Combine(BlocksFolderPath, encoder.EncodeData(hash.ToBytes()));
+				if (Directory.Exists(filePath))
 				{
-					var fileName = Path.GetFileName(filePath);
-					if (!encoder.IsValid(fileName))
+					try
 					{
-						Logger.LogTrace<WalletService>($"Filename is not a hash: {fileName}.");
-						continue;
+						var blockBytes = await File.ReadAllBytesAsync(filePath);
+						block =  Block.Load(blockBytes, Synchronizer.Network);
 					}
-
-					if (hash == new uint256(fileName))
+					catch (Exception)
 					{
-						var blockBytes = await File.ReadAllBytesAsync(filePath);
-						try
-						{
-							return Block.Load(blockBytes, Synchronizer.Network);
-						}
-						catch (Exception)
-						{
-							// In case the block file is corrupted we get an EndOfStreamException exception
-							// Ignore any error and continue by re-downloading the block.
-							break;
-						}
+						// In case the block file is corrupted and we get an EndOfStreamException exception
+						// Ignore any error and continue to re-downloading the block.
 					}
 				}
 			}
+
 			cancel.ThrowIfCancellationRequested();
```

Comment:
No need to throw here. The function already finished. Maybe throw outside the block and/or can't `ReadAllBytesAsync` get a cancellation token??